### DISCUSSION TITLE
How can I make Razzle not repeat itself infinity times when saving in VS Code a file that is already saved?

### DISCUSSION THREAD

**Discussion by silviubogan** [jaredpalmer/razzle#1638] [Q&A] [answered]
I begin with:

```shell
$ npm start
```

Razzle starts up, the browser can access the app. I have in VS Code a file inside the Razzle-based project (e.g. some .ts file) that I save again although it is already saved. I get this in the terminal:

![image](https://user-images.githubusercontent.com/198924/123179958-0d19ae80-d493-11eb-98fe-b8fe3875fada.png)

![image](https://user-images.githubusercontent.com/198924/123180031-2ae71380-d493-11eb-8196-d1c32d135b24.png)

It repeats itself until I press `Ctrl-C`.

My razzle.config.js is this:

```js
module.exports = {
  plugins: ["scss"],
  modifyWebpackOptions({ env: { target, dev }, options: { webpackOptions } }) {
    webpackOptions.notNodeExternalResMatch = (request, context) => {
      return /(react-syntax-highlighter|react-virtualized|react-select-virtualized|tiny-slider)/.test(
        request
      );
    };
    webpackOptions.babelRule.include = webpackOptions.babelRule.include.concat([
      /react-syntax-highlighter/,
      /react-virtualized/,
      /react-select-virtualized/,
      /tiny-slider/,
    ]);
    return webpackOptions;
  },
};
```

What can I do next? Is it something obvious or should I post the verbose debug output? If yes, all of it or just one of the flags inside `debug` in razzle.config.js? Could it be a file system issue related to the available watchers or something like this? If yes, then I have some more details to put here on request.

Thanks!

**Comment by fivethreeo** [comment]
You are probably watching some files in the build directory or something

**Reply by silviubogan** [reply]
I'll do a reboot then. Thanks.

**Reply by silviubogan** [reply]
It does the same. I think I'll try without VS Code after another reboot.

**Reply by fivethreeo** [reply]
You still have no emit in tsconfig?

**Reply by fivethreeo** [reply]
Tsc is only used for type checking. We transpile with babel. There is also the typescript plugin for razzle if you need to use ts-loader instead of babel.

**Reply by silviubogan** [reply]
> You still have no emit in tsconfig?

Yes.

**Reply by silviubogan** [reply]
> Tsc is only used for type checking. We transpile with babel. There is also the typescript plugin for razzle if you need to use ts-loader instead of babel.

Are you suggesting me to use the TS plugin for Razzle instead of the builtin support? I thought that all TS does is type checking plus some things that babel does for compatibility with older ES versions. Currently I do not see why it would be advantageous to use the TS plugin for Razzle.

**Reply by silviubogan** [reply]
> It does the same. I think I'll try without VS Code after another reboot.

I did this, I used just two terminal instances (and an open browser window), one with `npm start` and the other with NeoVim.

Pressing `:w` even though the file has not been changed, or at least not the contents of the file, produces this:

![image](https://user-images.githubusercontent.com/198924/123243440-b2607100-d4eb-11eb-81d2-e7e24b708418.png)

**Reply by silviubogan** [reply]
If I set `noEmit` to `false` in *tsconfig.json*, now I get, instead of 0, hundreds of errors. All of them [TS5055](https://stackoverflow.com/questions/40460790/getting-a-storm-of-error-ts5055-messages-on-every-typescript-compile-in-vs-cod#42119594), they are trying to output the `*.js` files over themselves, like the output path is not set:

> Cannot write file '/path/to/my/project/x/y/z/abc.js' because it would overwrite input file.

Do you think it is possible that this causes the autorepeating compilation process that I talked about in this discussion?

What is the best strategy to overcome this situation?

1. I could convert all the JS files to TS or TSX but it would take some time.
2. Or is it worth it to migrate to ts-loader? Does ts-loader provide a way to avoid this issue?

**Reply by silviubogan** [reply]
There are 148 errors of type TS5055 but the auto-repeating behavior is gone although I only changed `noEmit` to `false`.

**Comment by fivethreeo** [comment]
noEmit must be true

**Reply by fivethreeo** [reply]
https://webpack.js.org/configuration/watch/

**Reply by fivethreeo** [reply]
You can try setting options.verbose in razzle.

**Reply by fivethreeo** [reply]
https://github.com/jaredpalmer/razzle/blob/eb8c086a1efdd946a57bcc3c6d99d0b42446ea37/packages/razzle/config/createConfigAsync.js#L306

**Reply by silviubogan** [reply]
I don't see anything interesting with options.verbose set to true:

![image](https://user-images.githubusercontent.com/198924/123286008-33ccf900-d516-11eb-85c2-512c470fae62.png)

**Reply by fivethreeo** [reply]
https://webpack.js.org/configuration/watch/#vim

**Reply by silviubogan** [reply]
> https://github.com/jaredpalmer/razzle/blob/eb8c086a1efdd946a57bcc3c6d99d0b42446ea37/packages/razzle/config/createConfigAsync.js#L306

`webpackOptions.watchIgnorePaths` is `[ '/path/to/my/project/build/assets.json' ]` in my case.

**Reply by silviubogan** [reply]
> https://webpack.js.org/configuration/watch/#vim

I use VS Code usually. Do you think that VS Code does something similar to Vim?

**Reply by fivethreeo** [reply]
What is in your .env

**Reply by fivethreeo** [reply]
Which os is this?

**Reply by fivethreeo** [reply]
https://github.com/storybookjs/storybook/issues/14342#issuecomment-832764305

**Reply by silviubogan** [reply]
> What is in your .env

```
$ uname -a
Linux silviu-omenbyhplaptop17cb0xxx 5.12.9-1-MANJARO #1 SMP PREEMPT Thu Jun 3 14:56:42 UTC 2021 x86_64 GNU/Linux

$ cat .env
PORT=3005
PUBLIC_URL=http://localhost:3005/
PUBLIC_PATH=http://localhost:3005/
```

**Reply by fivethreeo** [reply]
I am out of ideas, but something with watching probably. Try a basic https://github.com/jaredpalmer/razzle/tree/master/examples/with-typescript - to see if it is  a razzle bug.

**Reply by silviubogan** [reply]
@fivethreeo  OK, thanks!

**Reply by silviubogan** [reply]
@fivethreeo I think the issue is related to [this](https://github.com/guard/listen/blob/master/README.md#increasing-the-amount-of-inotify-watchers). I've correctly (I've did this incorrectly before) used the method for ArchLinux since Manjaro Linux is based on ArchLinux and now I am rebooting and then testing again.

**Reply by fivethreeo** [reply]
https://gist.github.com/fivethreeo/0e2c6a467b6e1236b862082a317ca938

**Reply by fivethreeo** [reply]
This issue gives a warning about not enough watchers so is a different issue i think.

**Reply by silviubogan** [reply]
When starting npm start, within the first compilation process, the given plugin lists 114 JS files in the project, after the heading "The following files changed since last build:". Since I have 149 JS files in my project I think that the 114 files are the ones that are actually used, not just dead code.

Then I wait for both the client and the server to get compiled so that it says "App started http://localhost:3005".

I save an already saved file in VS Code and then it writes the same heading with the file that it thinks has changed below, in a list with a single item.

Then it starts the infinite loop:

![image](https://user-images.githubusercontent.com/198924/123411952-b5299780-d5b9-11eb-9492-137bc1ca8eca.png)

**Reply by fivethreeo** [reply]
I need to see those paths.

**Reply by silviubogan** [reply]
I masked them a little bit but I hope they are useful: https://gist.github.com/silviubogan/887d51d82bd2b7bad1244da2d22abd6b.

**Reply by fivethreeo** [reply]
And those are the same as shown on first run?

**Reply by silviubogan** [reply]
Yes.

**Reply by silviubogan** [reply]
I don't know if this is relevant, but typing `rs` inside the Razzle watch does not stop the infinite loop.

**Reply by fivethreeo** [reply]
Those files are js filed in the src dir or generated from ts?

**Reply by silviubogan** [reply]
JS files in the src dir.

**Reply by fivethreeo** [reply]
Strange it lists all as changed

**Reply by fivethreeo** [reply]
Webpack version?

**Reply by silviubogan** [reply]
4.44.2

**Reply by fivethreeo** [reply]
Maybe try 4.46.0? Not sure if it matters, but ..

**Reply by fivethreeo** [reply]
Any progress? I am thinking this could be a webpack config or a webpack plugin issue. Possibly some conflicting build dependencies. You use npm and resolutions, I would recommend yarn for resolutions, npm resolutions is a bit hackey. Not sure how you went about moving the app to razzle, the best would be to take the basic typescript example and add the code, then add deps with yarn and fix code till it builds.

**Reply by fivethreeo** [reply]
Or just let me at it 😛

**Reply by silviubogan** [reply]
Hi! I've converted all used .js files to .ts or .tsx inside the project and I am still getting the infinite-compilations-on-useless-save issue:

![image](https://user-images.githubusercontent.com/198924/123516070-f184f280-d6a2-11eb-9a1e-f0a244b54fd6.png)

**Reply by silviubogan** [reply]
I did not try yet with another WebPack version. I don't know which should I use, npm or yarn. I think npm is the oldest and it has worked for me very well in the past. I have to take another look at the Razzle examples.

> the best would be to take the basic typescript example and add the code, then add deps with yarn and fix code till it builds.

This I want to keep in mind. Thanks.

**Reply by silviubogan** [reply]
Actually, I missed some .js files. I'll keep you updated.

**Reply by silviubogan** [reply]
For me it works now that I have converted the last 2 .js files into .ts/x files. There still are a few .json files that don't cause problems. Even if I save them, they don't cause problems.

**Reply by fivethreeo** [reply]
So something touches all js files?

**Reply by silviubogan** [reply]
> So something touches all js files?

Maybe, I am not sure.

**Comment by fivethreeo** [comment, marked as answer]
Mixing js and ts seems to cause issues.

**Reply by fivethreeo** [reply]
If this is webpack 5.45+ this is a webpack bug. They are working on it.
